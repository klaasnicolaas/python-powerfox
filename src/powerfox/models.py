"""Asynchronous Python client for Powerfox."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass
class Device(DataClassORJSONMixin):
    """Object representing a Device from Powerfox."""

    device_id: str = field(metadata=field_options(alias="DeviceId"))
    date_added: datetime = field(
        metadata=field_options(
            alias="AccountAssociatedSince",
            deserialize=lambda x: datetime.fromtimestamp(x, tz=UTC),
        )
    )
    main_device: bool = field(metadata=field_options(alias="MainDevice"))
    bidirectional: bool = field(metadata=field_options(alias="Prosumer"))
    division: int = field(metadata=field_options(alias="Division"))
    name: str | None = field(metadata=field_options(alias="Name"), default=None)


# class Power

# class Heat

# class Water
