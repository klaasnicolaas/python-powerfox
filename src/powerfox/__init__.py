"""Asynchronous Python client for Powerfox."""

from .exceptions import (
    PowerfoxAuthenticationError,
    PowerfoxConnectionError,
    PowerfoxError,
)
from .models import Device, PowerMeter, Poweropti, WaterMeter, DeviceType
from .powerfox import Powerfox

__all__ = [
    "Device",
    "DeviceType",
    "PowerMeter",
    "Powerfox",
    "PowerfoxAuthenticationError",
    "PowerfoxConnectionError",
    "PowerfoxError",
    "Poweropti",
    "WaterMeter",
]
