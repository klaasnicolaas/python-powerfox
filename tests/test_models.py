"""Test the models for Powerfox."""

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from powerfox import (
    Device,
    DeviceType,
    HeatMeter,
    Powerfox,
    PowerMeter,
    Poweropti,
    WaterMeter,
)

from . import load_fixtures


async def test_all_devices_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    powerfox_client: Powerfox,
) -> None:
    """Test devices data function."""
    aresponses.add(
        "backend.powerfox.energy",
        "/api/2.0/my/all/devices",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("all_devices.json"),
        ),
    )
    devices: list[Device] = await powerfox_client.all_devices()
    assert devices == snapshot

    # Validate the human-readable property for each device
    for device in devices:
        assert isinstance(device.type, DeviceType)
        assert device.type.human_readable


async def test_power_meter_full_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    powerfox_client: Powerfox,
) -> None:
    """Test power meter full data function."""
    aresponses.add(
        "backend.powerfox.energy",
        "/api/2.0/my/power_device_id/current",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("power_meter_full.json"),
        ),
    )
    power_meter: Poweropti = await powerfox_client.device("power_device_id")
    assert power_meter == snapshot
    assert isinstance(power_meter, PowerMeter)


async def test_power_meter_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    powerfox_client: Powerfox,
) -> None:
    """Test power meter data function."""
    aresponses.add(
        "backend.powerfox.energy",
        "/api/2.0/my/power_device_id/current",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("power_meter.json"),
        ),
    )
    power_meter: Poweropti = await powerfox_client.device("power_device_id")
    assert power_meter == snapshot
    assert isinstance(power_meter, PowerMeter)
    assert power_meter.energy_usage_low_tariff is None
    assert power_meter.energy_usage_high_tariff is None


async def test_water_meter_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    powerfox_client: Powerfox,
) -> None:
    """Test water meter data function."""
    aresponses.add(
        "backend.powerfox.energy",
        "/api/2.0/my/water_meter_id/current",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("water_meter.json"),
        ),
    )
    water_meter: Poweropti = await powerfox_client.device("water_meter_id")
    assert water_meter == snapshot
    assert isinstance(water_meter, WaterMeter)


async def test_heat_meter_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    powerfox_client: Powerfox,
) -> None:
    """Test heat meter data function."""
    aresponses.add(
        "backend.powerfox.energy",
        "/api/2.0/my/heat_meter_id/current",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("heat_meter.json"),
        ),
    )
    heat_meter: Poweropti = await powerfox_client.device("heat_meter_id")
    assert heat_meter == snapshot
    assert isinstance(heat_meter, HeatMeter)


async def test_invalid_power_meter_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    powerfox_client: Powerfox,
) -> None:
    """Test invalid power meter data function."""
    aresponses.add(
        "backend.powerfox.energy",
        "/api/2.0/my/power_device_id/current",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("power_meter_invalid.json"),
        ),
    )
    power_meter = await powerfox_client.device("power_device_id")
    assert power_meter == snapshot
    assert isinstance(power_meter, PowerMeter)
    assert power_meter.energy_usage is None
    assert power_meter.energy_return is None
