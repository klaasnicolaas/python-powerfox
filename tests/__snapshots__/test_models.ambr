# serializer version: 1
# name: test_all_devices_data
  list([
    Device(device_id='9x9x1f12xx3x', date_added=datetime.datetime(2022, 10, 2, 9, 22, 35, tzinfo=datetime.timezone.utc), main_device=True, bidirectional=True, type=<DeviceType.POWER_METER: 0>, name='Wohnung unten'),
    Device(device_id='9x9x1f12xx2x', date_added=datetime.datetime(2022, 10, 2, 9, 22, 35, tzinfo=datetime.timezone.utc), main_device=True, bidirectional=True, type=<DeviceType.POWER_METER: 0>, name='Wohnung oben'),
    Device(device_id='9x9x1f12xx1x', date_added=datetime.datetime(2022, 10, 2, 9, 22, 35, tzinfo=datetime.timezone.utc), main_device=True, bidirectional=False, type=<DeviceType.COLD_WATER_METER: 1>, name=None),
  ])
# ---
# name: test_power_meter_data
  PowerMeter(outdated=False, timestamp=datetime.datetime(2023, 12, 20, 10, 48, 51, tzinfo=datetime.timezone.utc), power=111, energy_usage=1111.111, energy_return=111.111, energy_usage_high_tariff=None, energy_usage_low_tariff=None)
# ---
# name: test_power_meter_full_data
  PowerMeter(outdated=False, timestamp=datetime.datetime(2023, 12, 20, 10, 48, 51, tzinfo=datetime.timezone.utc), power=111, energy_usage=1111.111, energy_return=111.111, energy_usage_high_tariff=1111.111, energy_usage_low_tariff=0.011)
# ---
# name: test_water_meter_data
  WaterMeter(outdated=False, timestamp=datetime.datetime(2023, 12, 20, 10, 45, 7, tzinfo=datetime.timezone.utc), cold_water=1111.111, warm_water=0.0)
# ---
